### Project Introduction

Phishing is one of the most common cyberattacks used to steal sensitive information such as usernames, passwords, banking credentials, and personal data. Attackers create fraudulent websites that closely resemble legitimate websites to deceive users into revealing confidential information.

Traditional phishing detection techniques struggle to identify newly created or previously unseen phishing attacks.

Machine Learning provides a more adaptive approach by learning patterns from historical phishing and legitimate URLs. Instead of relying solely on predefined rules, machine learning algorithms analyse various URL, domain, DNS, SSL, and WHOIS-related characteristics to classify a website as either legitimate or phishing.

### Problem Statement

Phishing attacks continue to be one of the most significant cybersecurity threats, targeting individuals and organisations by impersonating legitimate websites to steal sensitive information. Detecting phishing websites manually is difficult due to the large number of URLs generated daily and the constantly evolving techniques used by attackers.

The objective of this project is to develop a machine learning model capable of automatically classifying a URL as either **Legitimate** or **Phishing** based on engineered URL, DNS, SSL, and domain-related features.

This is a **supervised binary classification** problem, where the model learns from labelled examples and predicts one of two possible classes:

- **Legitimate (0)** – A genuine and trusted website.
- **Phishing (1)** – A fraudulent website designed to deceive users.

The primary objective is to maximise the model's ability to correctly identify phishing websites while minimising false predictions. Multiple machine learning algorithms will be trained, evaluated, and compared using appropriate classification metrics to identify the best-performing model for phishing URL detection.

---

### 1. Dataset Information

#### Original Dataset

The foundation of this project is the **PhiUSIIL Phishing URL Dataset**, obtained from the UCI Machine Learning Repository. The original dataset contains labelled URLs together with a large number of pre-engineered features designed for phishing detection.

Rather than using the provided engineered features directly, only the **URL** and **Label** columns were retained. This decision was made to develop a completely custom feature engineering pipeline capable of extracting meaningful characteristics directly from raw URLs.

By starting with only the raw URL and its corresponding label, the project demonstrates the complete workflow of a real-world phishing detection system, where feature extraction is performed before machine learning model training. This approach provides greater flexibility, transparency, and a deeper understanding of the phishing detection process.

#### What was measured
What are the effective minimal sets of features that can be utilized in predicting phishing?

#### Why it was collected
With the rise in cybercrime, phishing remains a significant concern as it targets individuals with fake websites, causing victims to disclose their private information. The effective implementation of phishing detection relies on cost efficiency, with the increased feature extraction factor contributing to these costs. This advances ongoing efforts to improve the accuracy and efficiency of phishing detection mechanisms
and enhance cybersecurity defenses against malicious activities. This aims to explore various methods for detecting phishing emails using machine learning techniques by systematically evaluating and comparing different models.

#### How the data was obtained

Unlike many phishing detection projects that directly use pre-engineered features, this project generates a completely new machine learning dataset through a custom feature extraction pipeline.

The original PhiUSIIL dataset was used only as a source of labelled URLs. During preprocessing, only the **URL** and **Label** columns were retained, while all existing engineered features were discarded. For each URL, multiple custom Python modules automatically extracted lexical, domain, SSL, WHOIS, DNS, and internet-based characteristics.

The dataset generation pipeline consists of three specialised feature extraction modules:

- **Basic Features** – Extracts lexical properties directly from the URL, such as URL length, number of dots, special characters, path complexity, and HTTPS usage.
- **Domain Features** – Extracts domain-related characteristics including subdomains, suspicious keywords, brand impersonation indicators, top-level domain analysis, URL entropy, and redirect masking.
- **Internet Features** – Collects network-based information including SSL certificates, WHOIS records, domain age, DNS information, HTTP responses, webpage structure, HTML forms, password fields, and external hyperlink statistics.

The extracted features from each module are merged together and combined with the original class label to generate a new dataset named **Phishing_Features.csv**. This generated dataset forms the foundation for all subsequent stages of the project, including exploratory data analysis, feature selection, machine learning model development, and performance evaluation.


**What do the instances in this dataset represent?**
URLs and their corresponding webpages

**Has Missing Values?**
Yes


### Project Structure

To improve maintainability, readability, and scalability, the phishing URL detection system was organised into multiple Python modules, with each module responsible for a specific stage of the feature extraction process. This modular architecture follows good software engineering practices by separating feature extraction, dataset generation, and machine learning into independent components.

```
Phishing_URL_Detector/
│
├── Phishing_URL_Dataset.csv      # Original dataset
├── extraction.py                 # Main feature extraction pipeline
├── basic_features.py             # Lexical feature extraction
├── dmn_features.py               # Domain feature extraction
├── internet_features.py          # SSL, WHOIS, DNS and HTTP features
├── Phishing_Features.csv         # Generated feature dataset
├── phishing_url_detector.ipynb   # Machine learning notebook
└── README.md                     # Project documentation
```

Each module has a clearly defined responsibility. The original dataset serves as the source of labelled URLs, the feature extraction modules generate meaningful characteristics from each URL, and the generated dataset is subsequently used to train and evaluate multiple machine learning models.

### Feature Categories

The custom feature extraction pipeline generates features from multiple sources to capture different characteristics of phishing websites. Rather than relying on a single type of information, the extracted features are grouped into logical categories, providing a comprehensive representation of each URL.

| Feature Category | Description | Example Features |
|------------------|-------------|------------------|
| **Lexical Features** | Features extracted directly from the URL string without internet access. | URL length, domain length, path length, number of dots, digits, special characters, HTTPS usage, IP address detection. |
| **Domain Features** | Features describing the structure and characteristics of the domain name. | Subdomain count, suspicious keywords, brand spoofing, URL entropy, top-level domain analysis, redirect masking. |
| **Internet Features** | Features requiring communication with external services. | SSL certificate information, WHOIS records, domain age, DNS records, HTTP status codes. |
| **Webpage Features** | Features extracted after analysing the webpage content. | HTML forms, password fields, suspicious page titles, external hyperlink ratio. |

By combining lexical, domain, internet, and webpage-based features, the generated dataset provides a comprehensive representation of each URL, enabling machine learning models to distinguish between legitimate and phishing websites more effectively.

---

### Understanding URL Structure

Before extracting features from a URL, it is important to understand its structure. A **URL (Uniform Resource Locator)** is the address used to locate resources on the Internet. Every URL consists of multiple components, and many of the features used in phishing detection are derived directly from these components.

For example, features such as the URL length, domain name, subdomains, path length, number of query parameters, and the protocol (HTTP or HTTPS) are all obtained by analysing different parts of a URL.

Understanding these components will make it easier to understand the feature engineering process later in this notebook.

---

### Example URL

```
https://login.secure-paypal.com.account-verification.xyz:8080/security/update/login.php?id=10293&session=abc123#section2
```

Let's break this URL into its individual components.

---

####  Scheme (Protocol)

```
https://
```

The **scheme**, also known as the **protocol**, specifies how communication takes place between the client (your browser) and the web server.

Common protocols include:

| Protocol | Description |
|----------|-------------|
| HTTP | HyperText Transfer Protocol (not encrypted) |
| HTTPS | HyperText Transfer Protocol Secure (encrypted using SSL/TLS) |
| FTP | File Transfer Protocol |

Although HTTPS encrypts communication, attackers can also obtain SSL certificates. Therefore, HTTPS alone does **not** guarantee that a website is legitimate.

---

####  Host

```
login.secure-paypal.com.account-verification.xyz
```

The **host** identifies the server that hosts the website.

It consists of:

- Registered Domain
- Subdomains

##### Registered Domain

```
account-verification.xyz
```

The registered domain is the domain that has been officially registered.

It consists of:

- **Domain Name:** `account-verification`
- **Top-Level Domain (TLD):** `xyz`

Common Top-Level Domains include:

- `.com`
- `.org`
- `.net`
- `.edu`
- `.gov`

---

##### Subdomains

```
login.secure-paypal.com
```

Everything before the registered domain is considered the **subdomain**.

Subdomains are often used to organise different sections of a website.

For example:

```
mail.google.com
```

- Subdomain: `mail`
- Registered Domain: `google.com`

Phishing websites frequently create long and misleading subdomains to trick users into believing they are visiting a trusted website.

For example:

```
login.paypal.com.fake-domain.xyz
```

Although the URL appears to contain **paypal.com**, the actual registered domain is **fake-domain.xyz**.

---

### Port (Optional)

```
:8080
```

The **port number** specifies which network service on the server should receive the request.

Common ports include:

| Port | Service |
|------|----------|
| 80 | HTTP |
| 443 | HTTPS |
| 21 | FTP |
| 22 | SSH |
| 8080 | Alternative HTTP |

If no port is specified, the browser automatically uses the default port associated with the protocol.

---

### Path

```
/security/update/login.php
```

The **path** specifies the location of the requested resource on the web server.

It may point to:

- A webpage
- An image
- A file
- A script
- An API endpoint

Long or deeply nested paths may sometimes indicate suspicious behaviour.

---

### Query Parameters

```
?id=10293&session=abc123
```

Query parameters provide additional information to the server.

Each parameter consists of a **key-value pair**.

For this example:

| Parameter | Value |
|-----------|-------|
| id | 10293 |
| session | abc123 |

Multiple parameters are separated using the `&` character.

Large numbers of query parameters or unusually long query strings may indicate suspicious URLs.

---

### Fragment (Optional)

```
#section2
```

A **fragment** identifies a specific section within a webpage.

Fragments are processed by the browser and are generally **not sent to the web server**.

---

### Complete URL Breakdown

```
https://login.secure-paypal.com.account-verification.xyz:8080/security/update/login.php?id=10293&session=abc123#section2

│
├── Scheme (Protocol)
│   https
│
├── Host
│   ├── Subdomains
│   │   login.secure-paypal.com
│   │
│   ├── Registered Domain
│   │   account-verification.xyz
│   │
│   └── Top-Level Domain
│       xyz
│
├── Port
│   8080
│
├── Path
│   /security/update/login.php
│
├── Query Parameters
│   id=10293
│   session=abc123
│
└── Fragment
    section2
```

---

### Why is Understanding URL Structure Important?

Many of the features extracted in this project are derived directly from different components of a URL. Understanding the structure of a URL makes it easier to understand how each feature is generated and why it may help distinguish legitimate websites from phishing websites.

| URL Component | Example Features |
|--------------|------------------|
| Entire URL | URL length, number of digits, number of special characters |
| Scheme | Uses HTTPS |
| Registered Domain | Domain length, suspicious keywords, Top-Level Domain |
| Subdomains | Number of subdomains |
| Path | Path length, file extension |
| Query Parameters | Number of parameters, query length |

---

### Loading the Dataset

Before we can do anything with machine learning, we need to load our data, and one of the most common formats for storing data is CSV, which stands for *Comma-Separated Values*. A CSV file is simply a plain text file where each row represents a record (like a person, a review, or a product), and each value is separated by a comma.

We will load CSV files into a program.
 
Using the `pandas` library and its `read_csv()` function (the most flexible and beginner-friendly)

It has its own strengths:

*pandas* is the most popular option because it makes working with tables of data very easy. It automatically handles many of the common problems you might run into and gives you a powerful tool called a *DataFrame*, a table that you can search, filter, and analyse with just a few lines of code.

Few things to check when loading a CSV file:

- *Header row*: Does the first line of your file contain column names (like "Name", "Age", or "Review")? If so, most tools can detect and use it automatically. If not, you'll need to tell the tool what the column names should be.

- *Comments*: Some CSV files include comment lines that start with a symbol like `#`. These lines are meant for humans, not the program, so you may need to tell your loading function to skip them.

- *Delimiter*: Although CSV stands for "comma-separated", sometimes other characters are used instead, like tabs (`\t`) or semicolons (`;`). Make sure you specify the correct one so the data loads properly.

---

In [490]:
# Install required libraries
%pip install pandas


Note: you may need to restart the kernel to use updated packages.


In [491]:
import pandas as pd

# Load the Phishing Dataset
filename = "Phishing_Features.csv"

# Read the dataset while preserving valid category values
# such as "None". This prevents pandas from automatically
# converting them into missing values (NaN).

df = pd.read_csv(
    filename,
    keep_default_na=False
)

### Checking the Dataset

After loading a dataset, it is good practice to inspect the first few rows before performing any analysis. This helps confirm that the dataset has been loaded correctly and provides a quick overview of its structure.

The head() function displays the first five rows of a DataFrame by default. It is commonly used to verify that the column names have been read correctly and that the values appear under the correct features.

We will now display the first five rows of the dataset to confirm that the data has been loaded successfully.

---

In [492]:
# Display the First Five Rows
df.head()

,URL,url_length,domain_length,path_length,query_length,param_count,has_equal,has_ampersand,query_ratio,dots,...,domain_age_days,new_domain,dns_exists,dns_ip_count,http_status,title_suspicious,form_count,has_password_field,external_link_ratio,label
0,https://www.southbankmosaics.com,32,24,0,0,0,0,0,0.0,2,...,2353,0,1,1,200,0,1,0,0.027777777777777776,1
1,https://www.uni-mainz.de,24,16,0,0,0,0,0,0.0,2,...,None,None,1,1,200,0,0,0,1.0,1
2,https://www.voicefmradio.co.uk,30,22,0,0,0,0,0,0.0,3,...,5931,0,1,4,200,0,0,0,0.22115384615384615,1
3,https://www.sfnmjournal.com,27,19,0,0,0,0,0,0.0,2,...,7218,0,1,2,403,None,None,None,None,1
4,https://www.rewildingargentina.org,34,26,0,0,0,0,0,0.0,2,...,2553,0,1,2,200,0,1,0,0.7647058823529411,1


### Understanding the Output

The output displays the first five rows of the generated phishing feature dataset.

Each row represents one URL, while each column represents an extracted feature or the corresponding class label.

At this stage, we are simply verifying that the dataset has been loaded correctly. We are not yet analysing the feature values themselves.

---

### Dataset Dimensions

Before performing exploratory data analysis or building machine learning models, it is useful to determine the size of the dataset.

The size of a dataset is described by its **dimensions**, which consist of:

- **Rows** – The number of observations (URLs).
- **Columns** – The number of variables, including the extracted features and the target label.

Pandas provides the `shape` attribute to return the dimensions of a DataFrame as a tuple in the form:

```
(number_of_rows, number_of_columns)
```

Knowing the dimensions of the dataset helps verify that the dataset has been loaded completely and provides an overview of its size before further analysis.

---

In [493]:
# How big is our data
print(df.shape)

(235795, 58)


### Understanding the Output

The output is displayed as a tuple in the following format:

```
(number_of_rows, number_of_columns)
```

For example:

```
(235795, 58)
```

This means:

- The dataset contains **235,795 URLs (rows)**.
- The dataset contains **58 columns**, including the engineered features and the target label.

Each row represents a single URL processed by the feature extraction pipeline, while each column represents either an extracted feature or the corresponding class label.

---

### Exploring more details of the data

Beyond checking the first few and last few rows, pandas offers convenient methods to quickly summarise your dataset. For instance, if you want to view column data types and any non-null counts, you can use data.info(). This can be useful to spot missing values or confirm that columns are numeric:

---

In [494]:
# Checking data info
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 235795 entries, 0 to 235794
Data columns (total 58 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   URL                  235795 non-null  str    
 1   url_length           235795 non-null  int64  
 2   domain_length        235795 non-null  int64  
 3   path_length          235795 non-null  int64  
 4   query_length         235795 non-null  int64  
 5   param_count          235795 non-null  int64  
 6   has_equal            235795 non-null  int64  
 7   has_ampersand        235795 non-null  int64  
 8   query_ratio          235795 non-null  float64
 9   dots                 235795 non-null  int64  
 10  hyphens              235795 non-null  int64  
 11  underscores          235795 non-null  int64  
 12  letters_ratio        235795 non-null  float64
 13  digit_ratio          235795 non-null  float64
 14  special_ratio        235795 non-null  float64
 15  https                235795 

From `data.info()`, we get a quick overview of the dataset's structure. It tells us:

- How many rows are in the dataset  
- Which columns are present  
- How many non-null (non-missing) values are in each column  
- What data type each column uses (such as `int64` for whole numbers, `float64` for decimal numbers, or `object` for text)

This information is especially helpful for spotting issues early. For example:
- If a column has fewer non-null values than the total number of rows, it likely contains *missing data*.
- If a column that should contain numbers is listed as `object`, the values might be stored as text, possibly due to formatting issues (like commas or symbols).

Checking the results of `data.info()`, allows us to make sure the dataset is properly formatted and ready for analysis or model training.

---

###  Analysing , visualising and Transforming tabular data

### Introduction  

We will explore how to analyse and preprocess a numeric/tabular dataset to gain meaningful insights before applying machine learning models. Understanding your data is a crucial first step in any data science workflow, as it influences how well your models perform and generalise to unseen data.

We will cover three key aspects of data exploration and preparation:  

- *Describing the dataset*: we will generate statistical summaries, such as measures of central tendency (mean, median) and dispersion (variance, standard deviation), to understand the distribution of each feature. Additionally, we will examine missing values, outliers, and correlations between variables to identify patterns and potential preprocessing steps.  

- *Visualising the dataset*: data visualisation helps us uncover hidden relationships, trends, and anomalies. We will use histograms, box plots, scatter plots, and correlation heatmaps to visually inspect distributions, detect skewness, and understand feature interactions.    
**Visualisations will be introduced throughout the notebook whenever they help explain the data. This approach ensures that each visualisation has a clear purpose rather than being included unnecessarily.**

This process is essential for making informed decisions in machine learning, as the quality of your input data directly impacts model accuracy and reliability.

###  Describing the Dataset

#### Statistical summary

Descriptive statistics help us understand how each part of a dataset is spread out. When we run `.describe()` on a pandas DataFrame in Python, we get a helpful summary that includes:

- *Count*:  </br>
Tells us how many values are present (not missing) for each column. This helps us spot any gaps in the data.

- *Mean*:  </br>
The average value. This gives us a rough idea of what's "typical" for each column.

- *Standard Deviation*:</br>
Shows how much the values vary. A high number means the data is more spread out.

- *Minimum Value*:  </br>
The smallest number in the column. This can help us find any errors or outliers.

- *25th Percentile (Q1)*:  </br>
A quarter of the values fall below this point. It gives us a sense of the lower end of the data.

- *50th Percentile (Median/Q2)*:  </br>
The middle value when the data is sorted. This shows us the centre of the data.

- *75th Percentile (Q3)*:  </br>
Three-quarters of the data fall below this point. It helps us understand the upper range.

- *Maximum Value*:  </br>
The largest value in the column. This can highlight extreme values or mistakes.

These summaries make it easier to spot missing information, unusual results, or parts of the data that might need special attention, like features with a very wide range of values:

---

In [495]:
# Summary Statistics 
df.describe()

,url_length,domain_length,path_length,query_length,param_count,has_equal,has_ampersand,query_ratio,dots,hyphens,...,deep_subdomain,very_deep_subdomain,php_page,long_path,path_depth,url_encoded,entropy,redirect_masking,has_at_misuse,label
count,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,...,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000,235795.000000
mean,35.366954,21.470396,3.735325,2.212977,0.060841,0.023270,0.011951,0.011065,2.257143,0.350249,...,0.022129,0.007795,0.022414,0.025217,0.419746,0.003762,3.960896,0.006391,0.006349,0.571895
std,41.214346,9.150793,23.501818,24.880892,0.555041,0.150761,0.108666,0.080187,0.922678,1.521763,...,0.147105,0.087944,0.148025,0.156783,0.970301,0.061218,0.315386,0.079689,0.079426,0.494805
min,14.000000,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.308993,0.000000,0.000000,0.000000
25%,24.000000,16.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.763342,0.000000,0.000000,0.000000
50%,28.000000,20.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.929214,0.000000,0.000000,1.000000
75%,35.000000,24.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,2.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,4.100817,0.000000,0.000000,1.000000
max,6097.000000,110.000000,4248.000000,3391.000000,51.000000,1.000000,1.000000,0.976108,99.000000,286.000000,...,1.000000,1.000000,1.000000,1.000000,33.000000,1.000000,5.335261,1.000000,1.000000,1.000000


To explore the categorical features in the Diabetes Hospital Readmission dataset, we use the `describe()` function with the parameter `include=["object"]`. This tells pandas to generate summary statistics only for columns stored as the **object** data type, which in this dataset represents the categorical features.

The output of `describe()` is displayed using **`.T` (transpose)**, which swaps the rows and columns of the table. This does not change the data itself; it simply makes the output easier to read by displaying one feature per row.


In [496]:
# Summary Statistics for Categorical Features
df.describe(include=["object"]).T

/var/folders/6w/6jhfgq8s7hb_9__z7h3g3tcm0000gn/T/ipykernel_64746/4028559741.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include=["object"]).T


,count,unique,top,freq
URL,235795,235370,https://att-104164.weeblysite.com/,2
subdomain,235795,39657,www,161291
domain,235795,169415,web,5757
tld,235795,1402,com,112565
registered_domain,235795,175509,web.app,5754
has_ssl,235795,3,1,161119
ssl_issuer_known,235795,1,None,235795
domain_age_days,235795,12477,None,96391
new_domain,235795,3,0,137518
dns_exists,235795,2,1,187561


### Understanding the Output

Unlike numerical features, categorical features are summarised using different statistics because they represent labels or categories rather than continuous numerical values.

The output includes:

- **Count** – The number of non-null values in the feature.
- **Unique** – The number of distinct categories present in the feature.
- **Top** – The most frequently occurring category (mode).
- **Freq** – The number of times the most common category appears.

These statistics provide a quick overview of the categorical features and help us understand the distribution of different categories before performing data preprocessing or feature engineering.

It is important to interpret these statistics carefully. During feature extraction, unavailable values were intentionally stored as the literal string `"None"` rather than being converted to `NaN`. Furthermore, the dataset was loaded using `keep_default_na=False`, ensuring that `"None"` remains a valid category instead of being treated as a missing value.

As a result, placeholder values such as `"None"` are included when calculating the **Count**, **Unique**, **Top** and **Freq** statistics. In some features, `"None"` may even appear as the most frequently occurring category.

This design provides greater flexibility during preprocessing. Rather than immediately treating unavailable values as missing data, each feature can be analysed individually. For example, if a feature contains a very high proportion of `"None"` values, it may be removed. Conversely, if only a small proportion of values are `"None"`, they can be retained as a separate category because the absence of information itself may provide useful information for distinguishing phishing websites from legitimate ones.

A dedicated missing value analysis will therefore be performed later in the notebook to determine the most appropriate handling strategy for each feature.

---

### 2. Data Quality Assessment

Before building any machine learning model, it is essential to assess the quality of the dataset. A machine learning model learns patterns directly from the data it is given. Therefore, if the dataset contains missing values, duplicate records, inconsistent data types, placeholder values or features with little useful information, these issues can negatively affect the model's performance and reduce its ability to make accurate predictions on unseen data.

Data Quality Assessment is the process of examining the dataset to identify potential issues that may need to be addressed before feature engineering, preprocessing and model training. Detecting these problems early allows appropriate preprocessing techniques to be applied and helps ensure that the final dataset is reliable, consistent and suitable for machine learning.

The main aspects of data quality assessment include:

- Identifying missing or unavailable values.
- Detecting duplicate records.
- Checking data types and feature consistency.
- Identifying placeholder values used during feature extraction.
- Assessing whether any features contain too little useful information.
- Verifying that the extracted features have been generated correctly.

For the generated phishing feature dataset, this stage is particularly important because several internet-based features (such as WHOIS information, SSL certificate details, DNS records and webpage content) may not always be available. 

Placeholder values such as `"None"` are not automatically identified by functions such as `info()` or `describe()`. Instead, they require a dedicated analysis during preprocessing to determine the most appropriate handling strategy for each feature. Depending on their frequency, these values may either be retained as a separate category or indicate that a feature contains too little useful information and should be removed.

In the following sections, we will systematically assess the quality of the dataset before proceeding with feature engineering, preprocessing and machine learning model development.
### What will we examine?

To assess the quality of the dataset, we will investigate several important aspects of the extracted features before beginning preprocessing.

1. **Placeholder Value Analysis**  
   Identify features containing the placeholder value `"None"` and determine the proportion of unavailable information in each feature.

2. **Duplicate Record Analysis**  
   Check whether duplicate records exist within the dataset, as duplicate URLs or feature vectors may introduce bias into the machine learning models.

3. **Data Type Verification**  
   Verify that each feature has an appropriate data type and identify any features that may require conversion during preprocessing.

4. **Feature Quality Assessment**  
   Examine the characteristics of each feature, including the number of unique values, dominant categories and low-information features, to determine whether they are likely to contribute meaningful information to the machine learning models.

The outcome of this assessment will guide the preprocessing strategy by identifying features that require cleaning, transformation or removal before feature engineering, feature selection and machine learning model development.

---

### 2.1 Missing and Unavailable Value Analysis

Missing or unavailable values are one of the most common data quality issues encountered in real-world datasets. They can reduce the amount of information available to a machine learning model and may negatively affect prediction performance if they are not handled appropriately.

It is important to identify which features contain unavailable information and determine its extent. This allows informed preprocessing decisions to be made based on the characteristics of each feature rather than applying the same strategy across the entire dataset.

Unlike many datasets, where missing values are represented as `NaN` or placeholder symbols such as `?`, unavailable information in this phishing feature dataset was intentionally stored as the literal string `"None"` during feature extraction, ensuring that `"None"` remains a valid category instead of being automatically interpreted as a missing value by Pandas.

This design preserves information about why a value is unavailable and provides greater flexibility during preprocessing. Depending on how frequently `"None"` appears within a feature, it may either be retained as a separate category or indicate that the feature contains too little useful information and should be removed.

In this section, we will first identify the occurrence of `"None"` values in each feature and calculate both their count and percentage. This analysis will help determine the most appropriate preprocessing strategy for each feature before building the machine learning models.

### Counting Placeholder Values

Before deciding how unavailable information should be handled, we first need to determine how frequently it occurs within each feature.

In this section, we count the number of `"None"` values present in each feature. This provides an overview of how much unavailable information exists throughout the dataset and helps identify features that may require further preprocessing.

Features containing a large number of `"None"` values may provide limited information, whereas features containing only a small number of `"None"` values may still be useful and can potentially retain `"None"` as a separate category.

---

In [497]:
# Count Placeholder Values ("None")

# Count the number of occurrences of the placeholder
# value "None" in each feature.
# Count the number of "None" values in each feature.
# Count the number of "None" values in each feature.
none_counts = (df == "None").sum()

# Calculate the percentage of "None" values.
none_percentages = (none_counts / len(df)) * 100

# Create a summary table.
none_summary = pd.DataFrame({
    "None Count": none_counts,
    "None Percentage (%)": none_percentages
})

# Keep only the features that contain at least one "None" value.
none_summary = none_summary[none_summary["None Count"] > 0]

# Round the percentage values for better readability.
none_summary["None Percentage (%)"] = (
    none_summary["None Percentage (%)"]
    .round(2)
)

# Sort the table by percentage (highest first).
none_summary = none_summary.sort_values(
    by="None Percentage (%)",
    ascending=False
)

# Display the summary table.
none_summary

,None Count,None Percentage (%)
ssl_issuer_known,235795,100.00
external_link_ratio,107370,45.54
title_suspicious,107358,45.53
form_count,107358,45.53
has_password_field,107358,45.53
domain_age_days,96391,40.88
new_domain,96391,40.88
http_status,62232,26.39
has_ssl,62135,26.35
dns_exists,48234,20.46


### Understanding the Output

The table summarises the number and percentage of `"None"` values present in each feature.

Only features containing at least one `"None"` value are displayed. This allows us to focus on the features that contain unavailable information while excluding features that are complete.

The **None Count** column shows the total number of observations containing the placeholder value `"None"` for each feature, while **None Percentage (%)** indicates the proportion of unavailable values relative to the entire dataset.

Sorting the table by percentage makes it easier to identify features with a high proportion of unavailable information that may require special attention during preprocessing.

Several features contain unavailable values because they rely on information obtained from external sources during feature extraction, such as SSL certificates, WHOIS records, DNS lookups and webpage content.

Features with a very high percentage of `"None"` values may contribute little useful information and could be considered for removal during preprocessing. In contrast, features with a moderate proportion of `"None"` values may still contain valuable information, and retaining `"None"` as a separate category may improve the model's ability to distinguish between legitimate and phishing websites.

These findings will guide the preprocessing decisions made later in the notebook.

---

### 2.2 Duplicate Record Analysis

Duplicate records are common in real-world datasets and may arise due to repeated data collection, merging datasets from multiple sources or data processing errors.

If duplicate records are not identified and handled appropriately, they can bias the machine learning models by giving greater importance to certain observations. This may lead to overly optimistic evaluation results and reduce the model's ability to generalise to unseen data.

In this section, we examine the dataset to determine whether duplicate records are present. Identifying duplicate entries helps ensure that each observation contributes equally to the learning process.

In [498]:
# Duplicate Record Analysis

# Count the number of duplicate records
# in the dataset.

duplicate_count = df.duplicated().sum()
print(f"Number of duplicate records: {duplicate_count}")

Number of duplicate records: 187


### Understanding the Output

The output reports the total number of duplicate records present in the dataset.

A duplicate record is an observation in which every feature has exactly the same values as another observation.

For this dataset, **187 duplicate records** were identified. Considering the dataset contains over 235,000 observations, duplicate records represent only a very small proportion of the data.

The presence of duplicate records does not necessarily indicate an error. They may arise because the same URL appears multiple times in the original dataset or because different URLs generate identical feature values after feature extraction.

---

### 2.3 Data Type Verification

Before preprocessing the dataset, it is important to verify that each feature is stored using an appropriate data type.

Different machine learning algorithms require numerical input, while categorical features typically need to be encoded before model training. Incorrect data types may lead to errors during preprocessing or cause features to be treated incorrectly during analysis.

In this section, we examine the data type of every feature to identify numerical and categorical variables. This information will help determine the appropriate preprocessing techniques, such as encoding categorical features or converting features to more suitable data types where necessary.


In [499]:
# Data Type Summary
# Create a summary table showing the data type of each feature.

dtype_summary = pd.DataFrame({
    "Feature": df.columns,
    "Data Type": df.dtypes.values
})
dtype_summary

,Feature,Data Type
0,URL,str
1,url_length,int64
2,domain_length,int64
3,path_length,int64
4,query_length,int64
5,param_count,int64
6,has_equal,int64
7,has_ampersand,int64
8,query_ratio,float64
9,dots,int64


### Understanding the Output

The table displays the data type assigned to each feature in the dataset.

In Pandas, the most common data types include:

- **int64** – Integer numerical values.
- **float64** – Decimal numerical values.
- **object** – Text or categorical values.
- **bool** – Boolean values represented as `True` or `False`.

The data types provide an initial indication of how each feature will be processed. Numerical features may require scaling or normalisation, while categorical features generally require encoding before they can be used by machine learning algorithms.

---

### 2.4 Feature Quality Assessment

In addition to checking for missing values, duplicate records and data types, it is important to understand the characteristics of each feature.

One useful measure is the number of unique values contained within a feature. This provides insight into the amount of information a feature contributes and helps identify features that may require different preprocessing strategies.

For example, a feature containing only one unique value provides no discriminatory information and is unlikely to contribute to the machine learning models. Conversely, features with a very large number of unique values may require specialised encoding techniques or further investigation.

In this section, we calculate the number of unique values present in each feature to gain a better understanding of the dataset before preprocessing.

### Feature Summary

The previous analyses examined individual aspects of the dataset, including data types, placeholder values and the number of unique values. While each provides useful information independently, combining these characteristics into a single summary makes it easier to evaluate the quality of every feature.

The feature summary table provides an overview of each feature by displaying its data type, the number of unique values and the proportion of unavailable (`"None"`) values.

This consolidated view helps identify features that may require special preprocessing, such as low-information features, high-cardinality features or features containing a large proportion of unavailable information.

In [500]:
# Feature Summary
# Create a summary table for every feature.

feature_summary = pd.DataFrame({
    "Data Type": df.dtypes,
    "Unique Values": df.nunique(),
    "None Count": (df == "None").sum(),
    "None Percentage (%)": ((df == "None").sum() / len(df)) * 100
})

feature_summary["None Percentage (%)"] = (
    feature_summary["None Percentage (%)"]
    .round(2)
)

feature_summary

,Data Type,Unique Values,None Count,None Percentage (%)
URL,str,235370,0,0.00
url_length,int64,481,0,0.00
domain_length,int64,101,0,0.00
path_length,int64,213,0,0.00
query_length,int64,394,0,0.00
param_count,int64,20,0,0.00
has_equal,int64,2,0,0.00
has_ampersand,int64,2,0,0.00
query_ratio,float64,2405,0,0.00
dots,int64,27,0,0.00


### 3. Data Preprocessing

The initial data assessment provided an overview of the quality and characteristics of the extracted features. Based on those findings, we can now begin preparing the dataset for machine learning.

Data preprocessing is one of the most important stages of the machine learning pipeline. The objective is to improve the quality of the dataset by removing unnecessary information, correcting inconsistencies and transforming features into a format suitable for model training.

The preprocessing steps performed in this project are based on the findings from the initial data assessment rather than applying generic preprocessing techniques. This ensures that every preprocessing decision is justified by the characteristics of the dataset.

The preprocessing workflow consists of the following stages:

1. Removing duplicate records.
2. Handling features containing a large proportion of `"None"` values.
3. Categorising features into numerical and categorical groups.
4. Handling remaining `"None"` values.
5. Converting features to appropriate data types.
6. Encoding categorical features.
7. Scaling numerical features.
8. Selecting the most informative features for model development.

### 3.1 Duplicate Record Removal

During the initial data assessment, **187 duplicate records** were identified within the dataset.

Although duplicate records represent only a very small proportion of the dataset, they may introduce unnecessary bias by causing certain observations to appear multiple times during model training.


In [501]:
# Remove duplicate records.

df = df.drop_duplicates()
print(f"Dataset shape after removing duplicates: {df.shape}")
# Verify that all duplicate records have been removed.

df.duplicated().sum()

Dataset shape after removing duplicates: (235608, 58)


np.int64(0)

### 3.2 Handling Features with High Proportions of "None" Values

The initial feature assessment showed that several features contain unavailable values represented by the placeholder `"None"`. However, not all features should be treated in the same way.

Features containing a small or moderate proportion of `"None"` values may still provide valuable information, particularly when the absence of information is itself meaningful. For example, unavailable WHOIS information or unsuccessful DNS lookups may help distinguish phishing websites from legitimate ones.

In contrast, features containing an extremely high proportion of `"None"` values contribute very little useful information because the majority of observations contain no valid data. Such features increase the dimensionality of the dataset while providing limited predictive value.

Therefore, before handling the remaining `"None"` values, features containing little or no useful information will be identified and removed.

In [502]:
# Display features containing "None" values.
none_summary

,None Count,None Percentage (%)
ssl_issuer_known,235795,100.00
external_link_ratio,107370,45.54
title_suspicious,107358,45.53
form_count,107358,45.53
has_password_field,107358,45.53
domain_age_days,96391,40.88
new_domain,96391,40.88
http_status,62232,26.39
has_ssl,62135,26.35
dns_exists,48234,20.46


### Identifying Features for Removal

From the feature summary, we observe that the feature **`ssl_issuer_known`** contains `"None"` for every observation in the dataset.

Since no valid information is available for this feature, it provides no discriminatory value for distinguishing phishing websites from legitimate websites.

Consequently, this feature will be removed before continuing with the remaining preprocessing steps.

###  Removing High-Cardinality Text Features

The feature extraction process generated several informative numerical and binary features from the original URL components, including the domain, subdomain and top-level domain (TLD). These engineered features capture important characteristics such as the number of subdomains, suspicious TLDs, spoofed brands, path depth and other phishing-related patterns.

As a result, retaining the original textual features would provide little additional information while significantly increasing the dimensionality of the dataset. In addition, these features contain a very large number of unique values, which may encourage machine learning models to memorise specific domains rather than learn general patterns that distinguish phishing websites from legitimate ones.

Therefore, the original high-cardinality textual features (`URL`, `subdomain`, `domain`, `registered_domain` and `tld`) are removed before further preprocessing and model development.

In [503]:
# Remove features that are not useful for model training.

columns_to_remove = [
    "ssl_issuer_known",
    "URL",
    "subdomain",
    "domain",
    "registered_domain",
    "tld"
]

df = df.drop(columns=columns_to_remove)
print("Features removed successfully.")

Features removed successfully.


### Classifying Features

Based on the feature summary and the definition of each extracted feature, the dataset can be categorised according to the type of information represented by each feature.

The majority of the extracted features are already stored using appropriate numerical data types. However, a small number of numerical features are currently stored as text because unavailable values are represented by the placeholder `"None"`.

Similarly, several features represent categorical information, including binary indicators and textual categories. While binary features are already encoded as `0` and `1`, the remaining categorical features require appropriate handling before they can be used for machine learning.

Therefore, the features are grouped into the following categories:

- **Numerical Features** – Features representing measurable quantities such as counts, lengths, ratios or continuous values.

- **Categorical Features** – Features representing categories or textual information. This group also includes binary features that are already encoded numerically and therefore require no further encoding.

This classification is based on the intended meaning of each feature rather than its current storage format within the dataset and forms the basis for the remaining preprocessing steps.

In [504]:
# Numerical features currently stored as strings.

numerical_string_features = [
    "domain_age_days",
    "dns_ip_count",
    "http_status",
    "form_count",
    "external_link_ratio"
]

# Categorical features containing the placeholder value "None".

categorical_none_features = [
    "has_ssl",
    "new_domain",
    "dns_exists",
    "title_suspicious",
    "has_password_field"
]

# Display the identified feature groups.

print("Numerical features stored as strings:")
print(numerical_string_features)

print("\nCategorical features containing 'None':")
print(categorical_none_features)

Numerical features stored as strings:
['domain_age_days', 'dns_ip_count', 'http_status', 'form_count', 'external_link_ratio']

Categorical features containing 'None':
['has_ssl', 'new_domain', 'dns_exists', 'title_suspicious', 'has_password_field']


### Understanding the Output

The output identifies the features that require special preprocessing before the dataset can be used for machine learning.

Five numerical features (`domain_age_days`, `dns_ip_count`, `http_status`, `form_count` and `external_link_ratio`) are currently stored as strings because unavailable values are represented by the placeholder `"None"`. During preprocessing, these placeholder values will be replaced with missing values (`NaN`) before the features are converted to appropriate numerical data types.

The output also identifies five categorical features (`has_ssl`, `new_domain`, `dns_exists`, `title_suspicious` and `has_password_field`) that contain the placeholder `"None"`. Unlike numerical features, these values represent unavailable categorical information and will be retained as a separate category during preprocessing.

Identifying these feature groups in advance provides a clear preprocessing strategy and ensures that each feature is handled according to its intended data type before model development.

In [505]:
# Feature Summary
# Create a summary table for every feature.

feature_summary = pd.DataFrame({
    "Data Type": df.dtypes,
    "Unique Values": df.nunique(),
    "None Count": (df == "None").sum(),
    "None Percentage (%)": ((df == "None").sum() / len(df)) * 100
})

feature_summary["None Percentage (%)"] = (
    feature_summary["None Percentage (%)"]
    .round(2)
)

feature_summary

,Data Type,Unique Values,None Count,None Percentage (%)
url_length,int64,481,0,0.00
domain_length,int64,101,0,0.00
path_length,int64,213,0,0.00
query_length,int64,394,0,0.00
param_count,int64,20,0,0.00
has_equal,int64,2,0,0.00
has_ampersand,int64,2,0,0.00
query_ratio,float64,2405,0,0.00
dots,int64,27,0,0.00
hyphens,int64,36,0,0.00


### 3.3 Analysis of Categorical Features Containing `"None"`

Several internet-based features contain the placeholder value `"None"` because the required information could not be retrieved during feature extraction. Before performing any preprocessing, these features are analysed to understand their value distributions and determine whether they contain sufficient information for machine learning.

For each feature, the following characteristics are examined:

- Number of unique values
- Frequency of each value
- Percentage distribution
- Dominant value percentage
- Percentage of missing (`"None"`) values

This analysis helps identify low-information features and provides evidence for subsequent preprocessing decisions.

In [506]:
# Analyse categorical features containing "None".

categorical_analysis = []

for feature in categorical_none_features:

    # Count each unique value.
    value_counts = df[feature].value_counts(dropna=False)

    # Calculate percentage distribution.
    value_percentages = (
        df[feature]
        .value_counts(normalize=True, dropna=False)
        * 100
    ).round(2)

    # Identify the dominant value.
    dominant_value = value_counts.idxmax()
    dominant_percentage = value_percentages.max()

    categorical_analysis.append({
        "Feature": feature,
        "Unique Values": df[feature].nunique(),
        "None (%)": round((df[feature] == "None").mean() * 100, 2),
        "Dominant Value": dominant_value,
        "Dominant (%)": dominant_percentage
    })

categorical_analysis = pd.DataFrame(categorical_analysis)

categorical_analysis

,Feature,Unique Values,None (%),Dominant Value,Dominant (%)
0,has_ssl,3,26.35,1,68.33
1,new_domain,3,40.84,0,58.36
2,dns_exists,2,20.45,1,79.55
3,title_suspicious,3,45.49,0,53.98
4,has_password_field,3,45.49,0,51.90


### Understanding the Output

The table summarises the distribution of each categorical feature containing the placeholder value `"None"`.

For every feature, the number of unique values, percentage of missing (`"None"`) values, dominant value, and dominant value percentage are reported. These statistics provide an overview of feature variability and indicate whether a feature is heavily dominated by a single value.

Features with meaningful variation are likely to provide useful information for machine learning, whereas features that are almost entirely composed of one value may be considered low-information features and investigated further before preprocessing.

At this stage, no features are removed or converted. The results are used solely to guide the preprocessing decisions in the following sections.

### 3.4 Analysis of Numerical Features Stored as Strings

The following features represent numerical measurements but are currently stored as strings because they contain the placeholder value `"None"` for unavailable information.

Before converting these features into numerical data types, their distributions are analysed to verify that they contain valid numerical values and to understand the extent of missing information.

For each feature, the following statistics are calculated:

- Number of missing (`"None"`) values
- Number of unique values
- Minimum value
- Maximum value
- Mean
- Median
- Standard deviation

This analysis provides the justification for replacing `"None"` with `NaN` and converting these features to numerical data types during the preprocessing stage.

In [507]:
# Analyse numerical features currently stored as strings.

numerical_analysis = []

for feature in numerical_string_features:

    # Convert the feature to numeric for analysis only.
    numeric_feature = pd.to_numeric(
        df[feature].replace("None", pd.NA),
        errors="coerce"
    )

    numerical_analysis.append({
        "Feature": feature,
        "None (%)": round((df[feature] == "None").mean() * 100, 2),
        "Unique Values": numeric_feature.nunique(),
        "Minimum": numeric_feature.min(),
        "Maximum": numeric_feature.max(),
        "Mean": round(numeric_feature.mean(), 2),
        "Median": round(numeric_feature.median(), 2),
        "Standard Deviation": round(numeric_feature.std(), 2)
    })

numerical_analysis = pd.DataFrame(numerical_analysis)

numerical_analysis

,Feature,None (%),Unique Values,Minimum,Maximum,Mean,Median,Standard Deviation
0,domain_age_days,40.84,12476,-78.0,17001.0,6443.53,6386.00,3091.40
1,dns_ip_count,20.45,19,1.0,25.0,1.58,1.00,1.00
2,http_status,26.39,74,103.0,999.0,255.81,200.00,95.61
3,form_count,45.49,141,0.0,620.0,1.60,1.00,5.20
4,external_link_ratio,45.50,18573,0.0,1.0,0.25,0.13,0.29


### Understanding the Output

The table summarises the statistical properties of the numerical features currently stored as strings.

The missing (`"None"`) percentage indicates how frequently information could not be retrieved during feature extraction. The remaining statistics describe the numerical distributions after temporarily treating `"None"` as missing values.

These results confirm whether each feature contains meaningful numerical measurements and provide evidence that the features can be safely converted from strings to numerical data types after appropriate missing value handling.

No preprocessing is performed at this stage. The analysis is used solely to guide the data type conversion and missing value handling steps in the following sections.

---

In [508]:
# Display a comprehensive summary for each feature before preprocessing.
# This includes:
# - Current data type
# - Python value type
# - Placeholder ("None") count and percentage
# - Number of unique values
# - Top 10 most frequent values

all_features = categorical_none_features + numerical_string_features

for feature in all_features:

    print(f"Feature: {feature}")
    print(f"{'-' * 15}")

    # Display the current data type.
    print(f"Current Data Type : {df[feature].dtype}")

    # Display the Python value type.
    python_type = df[feature].map(type).iloc[0].__name__
    print(f"\nPython Value Type : {python_type}")

    # Calculate the placeholder ("None") count and percentage.
    placeholder_count = (df[feature] == "None").sum()
    placeholder_percentage = (placeholder_count / len(df)) * 100

    print(
        f"\nPlaceholder ('None') : "
        f"{placeholder_count:,} ({placeholder_percentage:.2f}%)"
    )

    # Display the number of unique values.
    unique_values = df[feature].nunique(dropna=False)
    print(f"\nUnique Values : {unique_values:,}")

    # Display the most frequent values.
    print("\nTop 10 Most Frequent Values:")

    display(
        df[feature]
        .value_counts(dropna=False)
        .head(10)
        .rename_axis("Value")
        .to_frame("Count")
    )

Feature: has_ssl
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 62,086 (26.35%)

Unique Values : 3

Top 10 Most Frequent Values:


,Count
Value,
1,160993
None,62086
0,12529


Feature: new_domain
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 96,222 (40.84%)

Unique Values : 3

Top 10 Most Frequent Values:


,Count
Value,
0,137500
None,96222
1,1886


Feature: dns_exists
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 48,186 (20.45%)

Unique Values : 2

Top 10 Most Frequent Values:


,Count
Value,
1,187422
None,48186


Feature: title_suspicious
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 107,187 (45.49%)

Unique Values : 3

Top 10 Most Frequent Values:


,Count
Value,
0,127177
None,107187
1,1244


Feature: has_password_field
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 107,187 (45.49%)

Unique Values : 3

Top 10 Most Frequent Values:


,Count
Value,
0,122283
None,107187
1,6138


Feature: domain_age_days
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 96,222 (40.84%)

Unique Values : 12,477

Top 10 Most Frequent Values:


,Count
Value,
None,96222
4960,1000
4959,657
4961,639
4958,552
4447,457
4446,414
4445,333
2721,316


Feature: dns_ip_count
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 48,186 (20.45%)

Unique Values : 20

Top 10 Most Frequent Values:


,Count
Value,
1,116781
2,51625
None,48186
4,9506
3,7069
5,831
6,751
8,619
10,82


Feature: http_status
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 62,183 (26.39%)

Unique Values : 75

Top 10 Most Frequent Values:


,Count
Value,
200,128132
None,62183
404,24310
403,11039
410,2496
500,2141
400,1392
530,935
503,731


Feature: form_count
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 107,187 (45.49%)

Unique Values : 142

Top 10 Most Frequent Values:


,Count
Value,
None,107187
0,51548
1,34836
2,21077
3,9572
4,4648
5,2308
6,1200
7,658


Feature: external_link_ratio
---------------
Current Data Type : str

Python Value Type : str

Placeholder ('None') : 107,199 (45.50%)

Unique Values : 18,575

Top 10 Most Frequent Values:


,Count
Value,
None,107199
0,10852
0.0,6012
1.0,3604
0.5,730
0.2,695
0.16666666666666666,665
0.3333333333333333,655
0.25,621


### 3.5 Preprocessing Decision Summary

The previous analyses verified the current representation, placeholder values, and value distributions of the selected categorical and numerical features. Based on these findings, an appropriate preprocessing strategy was determined for each feature before model development.

All analysed features contain meaningful information and will therefore be retained. However, different preprocessing techniques are required depending on whether a feature represents categorical or numerical data.

The following table summarises the preprocessing action that will be applied to each feature.

| Feature | Feature Type | Current Representation | Preprocessing Action | Decision |
|---------|--------------|------------------------|----------------------|----------|
| **has_ssl** | Categorical | `"0"`, `"1"`, `"None"` (stored as strings) | Encode `"0"` → `0`, `"1"` → `1`, `"None"` → `2` | Keep |
| **new_domain** | Categorical | `"0"`, `"1"`, `"None"` (stored as strings) | Encode `"0"` → `0`, `"1"` → `1`, `"None"` → `2` | Keep |
| **dns_exists** | Categorical | `"1"`, `"None"` (stored as strings) | Encode `"1"` → `1`, `"None"` → `2` | Keep |
| **title_suspicious** | Categorical | `"0"`, `"1"`, `"None"` (stored as strings) | Encode `"0"` → `0`, `"1"` → `1`, `"None"` → `2` | Keep |
| **has_password_field** | Categorical | `"0"`, `"1"`, `"None"` (stored as strings) | Encode `"0"` → `0`, `"1"` → `1`, `"None"` → `2` | Keep |
| **domain_age_days** | Numerical | Numeric values stored as strings with `"None"` | Replace `"None"` with `NaN`, then convert to a numeric data type | Keep |
| **dns_ip_count** | Numerical | Numeric values stored as strings with `"None"` | Replace `"None"` with `NaN`, then convert to a numeric data type | Keep |
| **http_status** | Numerical | Numeric values stored as strings with `"None"` | Replace `"None"` with `NaN`, then convert to a numeric data type | Keep |
| **form_count** | Numerical | Numeric values stored as strings with `"None"` | Replace `"None"` with `NaN`, then convert to a numeric data type | Keep |
| **external_link_ratio** | Numerical | Numeric values stored as strings with `"None"` | Replace `"None"` with `NaN`, then convert to a numeric data type | Keep |

### Understanding the Output

The preprocessing summary defines the transformation that will be applied to each feature before model training. Although all selected features are currently stored as string values, categorical and numerical features require different preprocessing strategies.

Categorical features will be encoded into numerical values while preserving the placeholder (`"None"`) as a separate category. Numerical features will first have the placeholder (`"None"`) replaced with `NaN` before being converted to an appropriate numeric data type.

Since all analysed features contain meaningful information, no additional features are removed during this stage. These preprocessing decisions prepare the dataset for subsequent missing value imputation, feature scaling, and machine learning model development.

---

### 3.6 Handle Placeholder Values (`"None"`)

The previous analysis confirmed that several numerical features are stored as strings and use the placeholder value (`"None"`) to represent unavailable information. Before these features can be converted to appropriate numeric data types, the placeholder must first be replaced with the standard missing value representation (`NaN`).

Only the numerical features are processed during this step because the placeholder (`"None"`) in categorical features represents a separate category rather than a missing numerical value. These categorical features will be encoded in a later preprocessing step.

In [509]:
# Replace the placeholder value ("None") with NaN
# in the numerical features only.

df[numerical_string_features] = (
    df[numerical_string_features]
    .replace("None", pd.NA)
)

# Display the number of missing values
# after replacing the placeholder.

display(
    df[numerical_string_features]
    .isna()
    .sum()
    .rename("Missing Values")
    .to_frame()
)

,Missing Values
domain_age_days,96222
dns_ip_count,48186
http_status,62183
form_count,107187
external_link_ratio,107199


In [510]:
# Verify that all placeholder values ("None")
# have been replaced with NaN.

remaining_placeholders = (
    df[numerical_string_features]
    .isin(["None"])
    .sum()
)

display(
    remaining_placeholders
    .rename("Remaining 'None'")
    .to_frame()
)

,Remaining 'None'
domain_age_days,0
dns_ip_count,0
http_status,0
form_count,0
external_link_ratio,0


### Understanding the Output

The output confirms that all placeholder values (`"None"`) within the numerical features have been successfully replaced with `NaN`. The reported missing value counts now represent actual missing values recognised by Pandas rather than string placeholders.

This transformation is required before converting the numerical features to appropriate numeric data types, as placeholder strings cannot be interpreted as numerical values. The categorical features remain unchanged because their placeholder values represent a valid category and will be encoded during a later preprocessing step.

### 3.7 Convert Numerical Features to Numeric Data Types

The previous preprocessing step replaced the placeholder value (`"None"`) with `NaN` for all numerical features. However, these features are still stored as strings and cannot be used directly for statistical analysis or machine learning.

This step converts each numerical feature to an appropriate numeric data type using Pandas. Any missing values represented by `NaN` are preserved during the conversion process and will be handled in a later preprocessing step.

In [511]:
# Convert the numerical features
# from strings to numeric data types.

df[numerical_string_features] = (
    df[numerical_string_features]
    .apply(pd.to_numeric)
)

# Display the updated data types.

display(
    df[numerical_string_features]
    .dtypes
    .rename("Data Type")
    .to_frame()
)

,Data Type
domain_age_days,float64
dns_ip_count,float64
http_status,float64
form_count,float64
external_link_ratio,float64


In [512]:
# Display summary statistics
# to verify the conversion.

display(
    df[numerical_string_features]
    .describe()
    .T
)

,count,mean,std,min,25%,50%,75%,max
domain_age_days,139386.0,6443.528080,3091.400347,-78.0,4174.000000,6386.000000,9156.750000,17001.0
dns_ip_count,187422.0,1.575418,0.999153,1.0,1.000000,1.000000,2.000000,25.0
http_status,173425.0,255.809981,95.611789,103.0,200.000000,200.000000,402.000000,999.0
form_count,128421.0,1.596273,5.202059,0.0,0.000000,1.000000,2.000000,620.0
external_link_ratio,128409.0,0.247347,0.293080,0.0,0.042169,0.126126,0.320755,1.0


### Understanding the Output

The output confirms that all selected numerical features have been successfully converted from string values to appropriate numeric data types. This conversion enables numerical operations such as descriptive statistics, missing value imputation, feature scaling, and machine learning model training.

The summary statistics also confirm that the converted features contain valid numerical values while preserving missing values as `NaN`. These features are now ready for the remaining preprocessing steps.

### 3.8 Encode Categorical Features Using One-Hot Encoding

The numerical features have now been converted into appropriate numeric data types. However, the categorical features are still stored as text values and cannot be used directly by most feature selection techniques.

To prepare these features for feature selection, One-Hot Encoding is applied. This encoding technique creates a separate binary feature for each category without introducing any artificial ordering between categories.

Unlike the numerical features, the placeholder value (`"None"`) is **not** treated as a missing value. Instead, it represents that the feature extraction process could not determine the feature value. Therefore, `"None"` is preserved as a valid category and encoded alongside the other categories.

In [513]:
# Create a copy of the cleaned dataset for feature selection.
feature_selection_df = df.copy()

# Apply One-Hot Encoding to the categorical features.
feature_selection_df = pd.get_dummies(
    feature_selection_df,
    columns=categorical_none_features,
    dtype=int
)

# Display the first five rows.
display(feature_selection_df.head())

,url_length,domain_length,path_length,query_length,param_count,has_equal,has_ampersand,query_ratio,dots,hyphens,...,new_domain_1,new_domain_None,dns_exists_1,dns_exists_None,title_suspicious_0,title_suspicious_1,title_suspicious_None,has_password_field_0,has_password_field_1,has_password_field_None
0,32,24,0,0,0,0,0,0.0,2,0,...,0,0,1,0,1,0,0,1,0,0
1,24,16,0,0,0,0,0,0.0,2,1,...,0,1,1,0,1,0,0,1,0,0
2,30,22,0,0,0,0,0,0.0,3,0,...,0,0,1,0,1,0,0,1,0,0
3,27,19,0,0,0,0,0,0.0,2,0,...,0,0,1,0,0,0,1,0,0,1
4,34,26,0,0,0,0,0,0.0,2,0,...,0,0,1,0,1,0,0,1,0,0


### Understanding the Output

The output confirms that the categorical features have been successfully transformed into binary numerical features using One-Hot Encoding.

Each original category has been replaced by a separate binary feature that contains either `1` (category present) or `0` (category absent). This transformation allows the categorical information to be used by feature selection algorithms without introducing any artificial numerical ordering.

The placeholder category (`"None"`) has also been preserved as an independent binary feature. This ensures that unsuccessful feature extraction attempts remain distinguishable from the other valid categories rather than being treated as missing data.

### 3.9 Handle Missing Numerical Values for Feature Selection

After encoding the categorical features, the feature selection dataset still contains missing values in several numerical features. These missing values originated from the placeholder value (`"None"`), which was previously converted to `NaN`.

Before applying feature selection techniques, the missing numerical values must be handled because most algorithms cannot operate on incomplete data.

Median imputation is applied to the numerical features because it is less sensitive to skewed distributions and outliers than mean imputation. The categorical features do not require imputation because they have already been transformed into complete binary features through One-Hot Encoding.

In [514]:
from sklearn.impute import SimpleImputer

# Create a median imputer.
median_imputer = SimpleImputer(strategy="median")

# Replace missing values in the numerical features.
feature_selection_df[numerical_string_features] = (
    median_imputer.fit_transform(
        feature_selection_df[numerical_string_features]
    )
)

# Verify that no missing values remain in the numerical features.
missing_values = (
    feature_selection_df[numerical_string_features]
    .isnull()
    .sum()
)

display(missing_values)

domain_age_days        0
dns_ip_count           0
http_status            0
form_count             0
external_link_ratio    0
dtype: int64

### Understanding the Output

The output displays the number of missing values remaining in each numerical feature after applying median imputation.

A value of `0` for every feature confirms that all missing numerical values have been successfully replaced with their respective median values.

At this stage, the feature selection dataset contains no missing values. The One-Hot encoded categorical features remain unchanged, while the numerical features are now complete and suitable for feature selection.

This dataset will be used as the base dataset for the three feature selection techniques applied in this study.

---

### 4. Feature Selection

Feature selection is the process of identifying the most informative features for predicting the target variable. Although the dataset has already been cleaned and preprocessed, not every feature contributes equally to the classification task. Some features may contain limited predictive information, while others may be redundant.

Selecting relevant features offers several advantages, including reducing model complexity, improving computational efficiency, minimizing overfitting, and potentially improving predictive performance.

To obtain a reliable assessment of feature importance, three complementary feature selection techniques are applied in this study:

- **Chi-Square Test** – measures the statistical association between each feature and the target variable.

- **Mutual Information** – evaluates how much information each feature provides about the target, including non-linear relationships.

- **Random Forest Feature Importance** – estimates feature importance based on the contribution of each feature to decision tree splits.

The feature rankings produced by these methods are compared to identify the most consistently informative features for the final machine learning models.

### 4.1 Prepare Datasets for Feature Selection

The three feature selection techniques used in this study require slightly different data representations.

The current dataset already satisfies the requirements of the Chi-Square test because all features are numerical and non-negative. This same unscaled dataset is also suitable for Mutual Information and Random Forest Feature Importance.

In addition, a standardized copy of the dataset is prepared for use by machine learning algorithms that benefit from feature scaling, such as Logistic Regression, Support Vector Machines (SVM), and K-Nearest Neighbours (KNN). Only the numerical features are standardized, while the One-Hot encoded categorical features remain unchanged because they already represent binary values.

As a result, two datasets are prepared:

- **`feature_selection_df`** – the original unscaled dataset used for feature selection.

- **`model_scaled_df`** – a standardized copy of the dataset that will be used later for machine learning models requiring feature scaling. from sklearn.preprocessing import StandardScaler




### Understanding the Output

The output presents the mean and standard deviation of each standardized numerical feature.

A mean close to **0** and a standard deviation close to **1** confirm that the numerical features have been successfully standardized using the StandardScaler.

Only the numerical features were standardized during this process. The One-Hot encoded categorical features remain unchanged because they already represent binary values (`0` and `1`).

Two datasets have now been prepared:

- **`feature_selection_df`** – the original unscaled dataset that will be used for feature selection.

- **`model_scaled_df`** – the standardized dataset that will be used later by machine learning algorithms that benefit from feature scaling, such as Logistic Regression, Support Vector Machine (SVM), and K-Nearest Neighbours (KNN).

Before applying the Chi-Square feature selection algorithm, the numerical features in the unscaled dataset are examined for negative values because the Chi-Square test requires all input features to be non-negative.

---

### 4.2 Verify Numerical Features for Negative Values

The Chi-Square feature selection algorithm can only be applied to non-negative feature values. Therefore, each numerical feature is examined to identify whether it contains any negative values.

If negative values are found, they are reviewed to determine whether they represent valid observations or values that require preprocessing before feature selection.


In [516]:
# Create an empty list to store
# information about features
# containing negative values.
negative_features = []

# Exclude the target variable.
predictors = (
    feature_selection_df
    .drop(columns=["label"])
)

# Check each numerical feature.
for feature in (
    predictors
    .select_dtypes(include=["number"])
    .columns
):

    # Count the number of
    # negative values.
    negative_count = (
        predictors[feature] < 0
    ).sum()

    # Store the feature if
    # negative values exist.
    if negative_count > 0:

        negative_features.append({
            "Feature": feature,
            "Minimum Value": (
                predictors[feature].min()
            ),
            "Negative Values": negative_count
        })

# Convert the results into
# a DataFrame.
negative_features_df = (
    pd.DataFrame(
        negative_features
    )
)

# Display the results.
display(negative_features_df)

,Feature,Minimum Value,Negative Values
0,domain_age_days,-78.0,1



### 4.3 Correct Invalid Negative Values

The previous verification identified that the **`domain_age_days`** feature contains negative values.

Since a domain age cannot be negative, these values are considered invalid. In addition, the Chi-Square feature selection algorithm requires all input features to be non-negative. Therefore, all negative values in the `domain_age_days` feature are replaced with **0** before feature selection.

In [517]:
# Replace all negative values
# in the domain_age_days
# feature with 0.
feature_selection_df[
    "domain_age_days"
] = (
    feature_selection_df[
        "domain_age_days"
    ].clip(lower=0)
)


### 4.4 Prepare the Dataset for Feature Selection

After correcting the invalid values, the dataset is separated into predictor variables and the target variable.

The predictor dataset (**`X`**) contains all input features, while the target dataset (**`y`**) contains the corresponding phishing URL labels. These datasets will be used as the input for the feature selection algorithms.


In [518]:
# Separate the predictor
# variables.
X = (
    feature_selection_df
    .drop(columns=["label"])
)

# Separate the target
# variable.
y = (
    feature_selection_df[
        "label"
    ]
)

# Display the dataset
# dimensions.
print(
    "Predictor Dataset Shape:",
    X.shape
)

print(
    "Target Dataset Shape:",
    y.shape
)

Predictor Dataset Shape: (235608, 60)
Target Dataset Shape: (235608,)


### 4.5 Chi-Square Feature Selection

The Chi-Square test is a filter-based feature selection technique that measures the statistical association between each feature and the target variable.

Features with larger Chi-Square scores have a stronger relationship with the target variable and are considered more informative for the classification task.

The Chi-Square test evaluates each feature independently and does not consider relationships between features. Therefore, its feature rankings will later be compared with those produced by Mutual Information and Random Forest Feature Importance to identify the most consistently informative features.

In [519]:
# Import the Chi-Square
# feature selection function.
from sklearn.feature_selection import chi2

# Compute the Chi-Square
# scores and corresponding
# p-values.
chi2_scores, chi2_pvalues = (
    chi2(X, y)
)

# Store the results.
chi2_results = (
    pd.DataFrame({
        "Feature": X.columns,
        "Chi-Square Score": chi2_scores,
        "P-Value": chi2_pvalues
    })
)

# Rank the features by
# their Chi-Square scores.
chi2_results = (
    chi2_results
    .sort_values(
        by="Chi-Square Score",
        ascending=False
    )
    .reset_index(drop=True)
)

# Display the ranked
# features.
display(
    chi2_results
)

,Feature,Chi-Square Score,P-Value
0,domain_age_days,1.417082e+07,0.000000e+00
1,path_length,1.175955e+06,0.000000e+00
2,http_status,7.590225e+05,0.000000e+00
3,query_length,6.982561e+05,0.000000e+00
4,url_length,5.891225e+05,0.000000e+00
5,path_depth,1.320783e+05,0.000000e+00
6,path_segments,7.931093e+04,0.000000e+00
7,hostname_length,7.341010e+04,0.000000e+00
8,domain_length,7.341010e+04,0.000000e+00
9,hyphens,6.404324e+04,0.000000e+00
